### PR TITLE
New Template for OPNsense on freebsd

### PR DESCRIPTION
**Pull request by benhopperau**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

**Comment by benhopperau** [comment]
The fail is for the date being in the future. This seems to be a timing thing, my timezone is GMT+10 so date is correct for me. However it looks like the test is for local time for the server, should the date be UTC? This might require an update to the contribution guide if thats the case.

If we re-submit the tests it will most likely pass.

**Comment by benhopperau** [comment]
Added some content to metadata.json to trigger CI to run tests again. Passed now that date no longer in the future for CI server.

File: opnsense-freebsd/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,368 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "vmSize": {
+      "type": "string",
+      "defaultValue": "Standard_A1_v2",
+      "allowedValues":[
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: remove allowedValues

File: opnsense-freebsd/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+      "type": "string",
+      "defaultValue": "opnsense-vm",
+      "metadata": {
+        "description": "Resource name of VM. This will not be the VM hostname."
+      }
+    },
+    "vmStorageAccountType": {
+      "type": "string",
+      "defaultValue": "Standard_LRS",
+      "allowedValues": [
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: remove allowedValues

File: opnsense-freebsd/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        "Standard_LRS",
+        "Standard_GRS",
+        "Standard_ZRS",
+        "Premium_LRS"
+      ],
+      "metadata": {
+        "description": "Type of VM storage account."
+      }
+    },
+    "networkPrefix": {
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: move all the networking settings to variables

File: opnsense-freebsd/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+        "description": "Type of public IP address."
+      }
+    },
+    "domainNamePrefix": {
+      "type": "string",
+      "metadata": {
+        "description": "DNS Label for the Public IP. Must be lowercase. It should match with the following regular expression: ^[a-z][a-z0-9-]{1,61}[a-z0-9]$ or it will raise an error."
+      }
+    },
+    "_artifactsLocation": {
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: need to add sasToken
See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#samples-that-contain-extra-artifacts-custom-scripts-nested-templates-etc) for how to handle artifacts

File: opnsense-freebsd/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+      "defaultValue": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/opnsense-freebsd/",
+      "metadata": {
+        "description": "The base URI where artifacts required by this template are located. When the template is deployed using the accompanying scripts, a private location in the subscription will be used and this value will be automatically generated."
+      }
+    }
+  },
+  "variables": {
+    "vmName": "[parameters('vmResourceName')]",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks', variables('vnetName'))]",
+    "adminUsername": "azureuser",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: this must be a parameter

File: opnsense-freebsd/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    "adminPassword": "Password1234",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: this must be a parameter

File: opnsense-freebsd/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+    "publicSubnet": "[concat(parameters('networkPrefix'), '-subnet-public')]",
+    "privateSubnet": "[concat(parameters('networkPrefix'), '-subnet-private')]",
+    "subnetPublicID": "[concat(variables('vnetID'),'/subnets/', variables('publicSubnet'))]",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: just use resourceID

resourceId('Microsoft.Network/virtualNetworks/subnets', variables('vnetName'), variables('publicSubnet'))

File: opnsense-freebsd/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+    "subnetPrivateID": "[concat(variables('vnetID'),'/subnets/', variables('privateSubnet'))]",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: resourceId('Microsoft.Network/virtualNetworks/subnets', variables('vnetName'), variables('privateSubnet'))

File: opnsense-freebsd/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+    "baseUriAzureCloud": "[parameters('_artifactsLocation')]",
+    "storageAccountName": "[concat('opnsense', uniquestring(resourceGroup().id))]",
+    "routeTableName": "[concat(parameters('networkPrefix'), '-routetable')]",
+    "nsgName": "[concat(parameters('networkPrefix'), '-nsg')]"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "name": "[variables('storageAccountName')]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: update apiVersion

File: opnsense-freebsd/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "accountType": "[parameters('vmStorageAccountType')]"
+      }
+    },
+    {
+      "type": "Microsoft.Network/routeTables",
+      "name": "[variables('routeTableName')]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: update apiVersion

File: opnsense-freebsd/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+              "addressPrefix": "0.0.0.0/0",
+              "nextHopType": "VirtualAppliance",
+              "nextHopIpAddress": "[parameters('privateNicIP')]"
+            }
+          }
+        ]
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: update apiVersion

File: opnsense-freebsd/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+              "routeTable": {
+                "id": "[resourceId('Microsoft.Network/routeTables', variables('routeTableName'))]"
+              }
+            }
+          }
+        ]
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: update apiVersion

File: opnsense-freebsd/azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "publicIPAllocationMethod": "[parameters('publicIPAddressType')]",
+        "dnsSettings": {
+          "domainNameLabel": "[parameters('domainNamePrefix')]"
+        }
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: update apiVersion

File: opnsense-freebsd/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+          }
+        ],
+        "enableIPForwarding": true,
+        "networkSecurityGroup": {
+          "id": "[resourceId('Microsoft.Network/networkSecurityGroups', variables('nsgName'))]"
+        }
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: update apiVersion

File: opnsense-freebsd/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+        "enableIPForwarding": true,
+        "networkSecurityGroup": {
+          "id": "[resourceId('Microsoft.Network/networkSecurityGroups', variables('nsgName'))]"
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Network/networkSecurityGroups",
+      "name": "[variables('nsgName')]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: update apiVersion

File: opnsense-freebsd/azuredeploy.json [hunk 16]
```diff
@@ -0,0 +1,368 @@
...
+              "access": "Allow",
+              "priority": 1000,
+              "direction": "Inbound"
+            }
+          }
+        ]
+      }
+    },
+    {
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: update apiVersion

File: opnsense-freebsd/azuredeploy.json [hunk 17]
```diff
@@ -0,0 +1,368 @@
...
+          "imageReference": {
+            "publisher": "MicrosoftOSTC",
+            "offer": "FreeBSD",
+            "sku": "11.1",
+            "version": "latest"
+          },
+          "osDisk": {
+            "name": "[concat(variables('vmName'),'-osdisk')]",
+            "createOption": "FromImage",
+            "vhd": {
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: use managed disks

File: opnsense-freebsd/azuredeploy.json [hunk 18]
```diff
@@ -0,0 +1,368 @@
...
+                "primary": false
+              },
+              "id": "[resourceId('Microsoft.Network/networkInterfaces', variables('privateNic'))]"
+            }
+          ]
+        },
+        "diagnosticsProfile": {
+          "bootDiagnostics": {
+            "enabled": "true",
+            "storageUri": "[concat('http://',variables('storageAccountName'),'.blob.core.windows.net')]"
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: use reference() to get the uri

File: opnsense-freebsd/azuredeploy.json [hunk 19]
```diff
@@ -0,0 +1,368 @@
...
+          }
+        }
+      }
+    },
+    {
+      "type": "Microsoft.Compute/virtualMachines/extensions",
+      "name": "[concat(variables('vmName'),'/configureopnsense')]",
+      "apiVersion": "2015-06-15",
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.json]: update apiVersion

File: opnsense-freebsd/azuredeploy.parameters.json [hunk 20]
```diff
@@ -0,0 +1,39 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
```
Comment (bmoore-msft) [inline opnsense-freebsd/azuredeploy.parameters.json]: remove all non-GEN values and use defaultValues in the template

**Comment by bmoore-msft** [comment]
Closing abandoned request - please reopen if you can address the feedback.

**Pull request closed without merging**